Feature/issue 30

stae1102 opened the pull request:
# 해당 이슈
#30

# 구현 기능

1. 특정 작성자의 게시물 조회
> GET /boards/search/:boardId

2. 내 게시물 조회
> GET /boards/search/:boardId

# 수정

1. 게시물 아이디를 통한 조회에서 파라미터를 사용하는데 쿼리를 받았던 문제점을 수정

# 제안 사항

게시물 조회 기능이 여러 개라서 경로 설정이 어려웠습니다. 조금 더 간결하고 명확하게 작성할 방법이 있을까요?

파라미터를 쓰면 어쩔 수 없이 그 경로는 뒤에 경로를 더 붙이거나 해야 해서 어렵네요.

stae1102 committed:
build: 추가 boards 모델 createdAt 컬럼

생성시간별로 게시물을 정렬하기 위해
생성시간 컬럼을 추가했습니다.

Fixes: #30

stae1102 committed:
perf: 수정 타입별 게시물 조회 반환값

타입별 게시물 조회를 최신순으로 정렬하였습니다.
새로 컬럼을 생성하였으므로, 이전 레코드들은 다 같은
시간대로 저장되었습니다.

Fixes: #30

stae1102 committed:
feat: 추가 유저 이름으로 게시물 조회

유저 이름을 사용하여 게시물을 조회하는 기능을 구현했습니다.
클라이언트는 유저의 아이디 값을 알 수 없으므로 닉네임과 비슷한
사용으로 이름을 통해 검색하도록 했습니다.

Fixes: #30

stae1102 committed:
chore: 수정 api 경로

경로를 조금 더 간결하게 search를 빼주었습니다.

Fixes: #30

stae1102 committed:
fix: 수정 /boards/search/:boardId 파라미터 사용

경로상에서 파라미터를 사용하지만 쿼리를 사용한 부분을
파라미터를 사용하도록 변경하였습니다.

Fixes: #30

stae1102 committed:
feat: 추가 boards 유저 이름으로 게시물 조회

유저 이름을 통해서 게시물을 조회하는 서비스를 만들었습니다.
파라미터에 이름을 기재하면 이를 통해서 해당 이름의 작성자를 찾아
모든 게시물을 최신순으로 반환합니다.

Fixes: #30

stae1102 committed:
feat: 추가 boards 내 모든 게시물 조회

User 데코레이터로 요청객체에 담긴 유저 정보를 사용하여
내가 작성한 게시물을 모두 가져옵니다.

Resolves: #30

stae1102 committed:
feat: 추가 boards 내 게시물 조회 컨트롤러

서비스만 커밋하여 컨트롤러도 커밋

Resolves: #30

stae1102 committed:
comment: 추가 boards controller, service 주석

controller
- 내 게시물, 특정 사용자 게시물 조회
service
- 내 게시물, 특정 사용자 게시물 조회

Ref: #30

stae1102 committed:
comment: 수정 boards controller 주석

api 경로가 잘못 기대되어 있어서 수정
GET /boards/search/:boardId
GET /boards/search/:name/boards

Ref: #30

stae1102 committed:
Merge branch 'develop' into feature/issue-30

stae1102 committed:
chore: 수정 누락된 임포트

MisterRuby commented:
```
GET /boards?keyword=검색어&category=작성자명&page=페이지번호

@Get()
getBoards(@Query() searchDto:SearchDto) {
    // SearchDto 를 통해 검색어, 검색 종류(작성자명, 게시글 등의 카테고리), 페이지 번호 등을 받음
}
```

하나의 컨트롤러로 처리할려면 쿼리 파라미터를 통해 처리하면 될 것 같습니다.
(URL 경로 중복 문제는 일단 리펙토링쪽에 의견을 추가했습니다. 이 부분은 같이 고민을 해봐야 될 것 같습니다.)

stae1102 committed:
feat: 변경 사용자 이름 및 게시물 종류를 통한 조회

경로가 너무 많아져 리소스가 너무 많아졌습니다. 그래서
기존에 사용자 이름 및 게시물 종류를 통한 조회 기능을
하나의 경로에서 받으며, 추가적으로 키워드를 통한 조회를 구현했습니다.
해당 라우트에서 사용할 수 있는 쿼리 및 반환값은 다음과 같습니다.
- page: 조회할 페이지 -> 각 페이지마다의 값을 반환
- kind: 게시물 종류 -> 게시물 종류에 따라 게시물을 반환
- name: 사용자 이름 -> 특정  작성자의 게시물을 반환
- keyword: 검색 키워드 -> 제목 혹은 내용에 키워드가 있을 시
그 게시물들을 반환
💡 모두 최신순으로 반환합니다.

Resolves: #30

stae1102 committed:
chore: 수정 코드 컨벤션 및 쿼리러너

카멜케이스가 적용되지 않은 변수를 수정했으며
쿼리러너를 사용하지 않는 부분을 수정했습니다.

Resolves: #30

stae1102 commented:
@MisterRuby
용범님 피드백 살려서 아래와 같이 한 번에 정리했고, 추가적으로 키워드를 사용한 조회도 포함했습니다~

```ts
📃boards.controller.ts
/**
   * 게시물 종류, 작성자 이름, 키워드(제목 혹은 내용)에 따라서 게시물을 검색하여 조회합니다.
   * @api GET /board?query=value[&query=value]
   * @query query { kind: 게시물 종류, name: 작성자 이름, keyword: 제목 및 내용 검색에 쓰일 단어}
   * @returns 각각의 용도에 따라 알맞게 게시물을 최신순으로 정렬하여 반환합니다.
   */
  @Get()
  searchBoards(
    @Query()
    query: {
      page: number;
      kind?: BoardKind;
      name?: string;
      keyword?: string;
    },
  ) {
    return this.boardsService.searchBoards(
      query.page,
      query.kind,
      query.name,
      query.keyword,
    );
  }
```

```typescript
📃boards.service.ts
async searchBoards(
    page: number,
    kind?: BoardKind,
    name?: string,
    keyword?: string,
  ) {
    const queryBuilder = this.dataSource
      .createQueryBuilder()
      .select('boards')
      .from(Boards, 'boards')
      .leftJoinAndSelect('boards.Author', 'author')
      .where('1 = 1')
      .take(30)
      .skip(30 * (page - 1))
      .orderBy('boards.createdAt', 'DESC');

    if (kind) {
      console.log('kind', kind);
      queryBuilder.andWhere('boards.kind = :kind', { kind });
    }
    if (name) {
      console.log('name', name);
      queryBuilder.andWhere('author.name = :name', { name });
    }
    if (keyword) {
      console.log('keyword', keyword);
      queryBuilder.andWhere(
        new Brackets((qb) => {
          qb.where('boards.title like :keyword', { keyword: `%${keyword}%` }),
            qb.orWhere('boards.content like :keyword', {
              keyword: `%${keyword}%`,
            });
        }),
      );
    }
    const data = await queryBuilder.getMany();
    return data;
  }
```

# 좋은 피드백 감사합니다 ㅎ_ㅎ

loveAlakazam commented:
오 페이지네이션까지 한거에요? 대박입니다 😎

MisterRuby commented:
```
async searchBoards(
    page: number,
    kind?: BoardKind,
    name?: string = '',
    keyword?: string = '',
  ) {
    const take = 30;

    const queryBuilder = this.dataSource
      .createQueryBuilder()
      .select('boards')
      .from(Boards, 'boards')
      .leftJoinAndSelect('boards.Author', 'author')
      .where('author.name LIKE :name', { name: `%{name}%` })
      .andWhere(
        new Brackets((qb) => {
          qb.where('boards.title like :keyword', { keyword: `%${keyword}%` }),
            qb.orWhere('boards.content like :keyword', {
              keyword: `%${keyword}%`,
            });
        }),
      );
      .take(take)
      .skip(take * (page - 1))
      .orderBy('boards.createdAt', 'DESC');

    if (kind) {
      console.log('kind', kind);
      queryBuilder.andWhere('boards.kind = :kind', { kind });
    }

    return queryBuilder.getMany();
  }
```
name 과 keyword 의 값이 없을 시 기본값을 빈문자열('') 로 지정하면 if로 분리하지 않고 처리할 수 있을 것 같습니다.
그리고 take 에 입력한 값은 별도로 변수로 선언하여 관리하는 게 좋아보입니다.
비즈니스 상 변경될 여지가 있을 수 있는 값이고 매직넘버는 지양해야 한다는 점에서도 그렇습니다.

새벽까지 수고하셨습니다!

stae1102 committed:
refactor: 수정 게시물 검색 기능

기존에 검색을 if문으로 두었는데, 이를 기본값으로
대체하여서 조금 더 간결하고 명확하게 수정했습니다.

Resolves: #30

stae1102 commented:
네 언급해주신 내용 수정했습니다 감사합니다!

이상 없으면 머지하겠습니다~

eunkangchoi committed:
Merge branch 'develop' into feature/issue-30
